api/src/main/java/org/pmiops/workbench/api/WorkspacesController.java

```diff
@@ -711,6 +712,15 @@ private FileDetail copyNotebookImpl(
     return ResponseEntity.ok(fileDetail);
   }
 
+  @Override
+  public ResponseEntity<KernelTypeResponse> getNotebookKernel(
+      String workspace, String workspaceName, String notebookName) {
+    return ResponseEntity.ok(
```

Comment:
it's the first call in the `getNotebookKernel()` function. This is something I'm a bit split on... on one hand, it's nice to check what authentication is needed for a controller method by looking at its first few lines but I think it also makes sense for the authentication check to be close to the code where the access is being performed. The latter approach is more bug free IMO if multiple services are looking to use that functionality since the access check comes with it and the developer doesn't need to bring up access checks to the controller method when importing functionality from services.

Reply:
I think our codebase mostly follows the controller-check pattern right now (immediately check auth in the method), and typically the service is not the one enforcing auth - I prefer this because it's very easy to verify by inspection. We could potentially move to a model where we push these checks lower, but I'm not sure I agree this is better. Putting the check lower perhaps means tighter enforcement but it also means potential redundancy.

That said, the patterns in this controller seems to be pretty mixed. I know there were some discussions around what belongs in a service vs controller that we never really reached consensus on - looks like half of this controller was refactored to reflect one ideology. If you feel strongly on this point, I would at least have a comment here and/or rename the method to make the auth check clearer.

Reply:
I agree about the potential (almost certain) redundancy if we apply this pattern everywhere. We'd need to implement some software pattern like caching ACLs for the context of a request so it doesn't get out of hand. The implementation might be tricky to get right (we'd have to clear the cache if we make a call to update ACLs), but I don't think the approach of caching an ACL is fundamentally less secure given that our current approach is to check once at the beginning of every API call. We would just be exposing a way to fetch that value.

Reply:
Regarding these changes though, I can move the access check to the controller method since it's our more adopted pattern.